### PR TITLE
Business pay

### PR DESCRIPTION
**Pull request by thorwolpert**
*Issue #:* n/a

*Description of changes:*
- initial business-pay service
- adding nats_queue
- adding gcp_queue
- adding gcp auth
- async support in Flask
- Python capped to v3.9 due to Queue dependencies

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: queue_services/business-pay/src/business_pay/resources/pay_filer.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    id: Optional[str] = None
+    status_code: Optional[str] = None
+    filing_identifier: Optional[str] = None
+    corp_type_code: Optional[str] = None
+
+
+def get_payment_token(ce: SimpleCloudEvent):
+    """Return a PaymentToken if enclosed in the cloud event."""
+    if (
+        (ce.type == "payment")
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/resources/pay_filer.py:166]: This needs to change to https://github.com/bcgov/sbc-common-components/blob/5a8d762cc17b6cc5113f42f775870e52f84e831a/python/src/sbc_common_components/utils/enums.py#L45

File: queue_services/business-pay/src/business_pay/services/gcp_auth.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+def verify_gcp_jwt(flask_request):
+    """Verify the bearer token as sign by gcp oauth."""
+    try:
+        bearer_token = flask_request.headers.get("Authorization")
+        token = bearer_token.split(" ")[1]
+        audience = current_app.config.get("GOOGLE_CLIENT_ID")
+        claim = id_token.verify_oauth2_token(
+            token, requests.Request(), audience=audience
+        )
+        return claim
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: Probably want to verify emails, that's what they did on their spec at least. 

https://cloud.google.com/pubsub/docs/authenticate-push-subscriptions

image

**Review by seeker25 — COMMENTED** [review]
Small tweaks

File: queue_services/business-pay/src/business_pay/resources/__init__.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+from .pay_filer import bp as pay_filer_endpoint
+
+
+def register_endpoints(app: Flask):
+    """Register endpoints with the flask application"""
+    # Allow base route to match with, and without a trailing slash
+    app.url_map.strict_slashes = False
+
+    app.register_blueprint(
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/resources/__init__.py:49]: do you need the OPS endpoints to pass the OCP health checks?

File: queue_services/business-pay/src/business_pay/services/gcp_auth.py (hunk 2 above)
Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: It's anote to check, they don't actually check and neither does namex from what I saw.

File: queue_services/business-pay/src/business_pay/resources/pay_filer.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        and (data := ce.data)
+        and isinstance(data, dict)
+        and (payment_token := data.get("paymentToken", {}))
```
[the remark this answers is not in the document]
Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/resources/pay_filer.py:169]: Ah, I had asked if there were changes from what it was last June.
Looks like the flow copied into NameX changed from this one.
I'll have a look.

File: queue_services/business-pay/src/business_pay/resources/pay_filer.py (hunk 1 above)
Comment (seeker25) [reply queue_services/business-pay/src/business_pay/resources/pay_filer.py:166]: The message needs to be `bc.registry.payment`

File: queue_services/business-pay/src/business_pay/services/gcp_auth.py (hunk 2 above)
Comment (seeker25) [reply queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: Names checks.. I looked at the PR today and they do

Comment (seeker25) [reply queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: https://github.com/bcgov/namex/blob/22d209d77bb982f20d407e685bb9eb2918478694/services/pubsub/gcp_queue/gcp_auth.py#L28

Comment (bolyachevets) [reply queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: Screenshot 2024-05-13 at 8 27 27 AM
did add the email check a week or two ago

Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: Yep, @seeker25 already put the link in.
I'll check.

Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/services/gcp_auth.py:53]: We have too much code copy/pasted and upgraded.
I'll add this to the gcp_queue package ... as soon as I get access back :(

File: queue_services/business-pay/src/business_pay/services/__init__.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+def create_filing_msg(identifier):
+    """Create the filing payload."""
+    filing_msg = {"filing": {"id": identifier}}
+    return filing_msg
+
+
+def create_email_msg(identifier, filing_type):
+    """Create the email message payload."""
+    email_msg = {"email": {"filingId": identifier, "type": filing_type}}
+    return email_msg
```
Comment (argush3) [inline queue_services/business-pay/src/business_pay/services/__init__.py:53]: I think we need to add `option`?  Emailer looks at that.

create_email_msg in queue common looks like this right now

https://github.com/bcgov/lear/blob/d6f8e6ec24e1603d39720f85205a6321dc441be3/queue_services/common/src/entity_queue_common/messages.py#L62-L65

Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/services/__init__.py:53]: I think that's only used by the names processor? 
Names has been moved to their own emailer, so this shouldn't receive those anymore.
I'll look more

Comment (argush3) [reply queue_services/business-pay/src/business_pay/services/__init__.py:53]: Think you'll need it even for non-names stuff.  Took a closer look.

image

image

Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/services/__init__.py:53]: ah, was only looking at the sigs ... 4 of them, sigh

**Comment by thorwolpert** [comment]
had to repoint the gcp_queue and logging to my own repo as move sbc-connect-common to bcgov broke the link.
will need TODO that back after syncing up the broken git chain

**Review by argush3 — APPROVED** [review]

File: queue_services/business-pay/devops/vaults.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    },
+    {
+        "vault": "entity",
+        "application": [
+            "postgres-legal",
+            "entity-pay",
+            "nats-pay",
+            "sentry"
+        ]
+    }
```
Comment (seeker25) [inline queue_services/business-pay/devops/vaults.json:16]: {
        "vault": "gcp-queue",
        "application": [
            "a083gt-dev",
            "topics"... etc.
        ]
    }

**Review by bolyachevets — APPROVED** [review]

File: queue_services/business-pay/src/business_pay/config.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        "subject": os.getenv("NATS_EMAILER_SUBJECT", "entity.email"),
+    }
+
+    ENTITY_FILER_TOPIC = os.getenv("ENTITY_FILER_TOPIC", "filer")
+    ENTITY_MAILER_TOPIC = os.getenv("ENTITY_FILER_TOPIC", "mailer")
+
+    ENVIRONMENT = os.getenv("ENVIRONMENT", "prod")
+
+    GOOGLE_CLIENT_ID = os.getenv("GOOGLE_CLIENT_ID", "google_client_id")
+    GCP_AUTH_KEY = os.getenv("GCP_AUTH_KEY", None)
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/config.py:134]: `GCP_AUTH_KEY = os.getenv("BUSINESS_GCP_AUTH_KEY`", None)`

File: queue_services/business-pay/src/business_pay/config.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    EMAILER_TOPIC = os.getenv("EMAILER_TOPIC", "namex-receipt-dev")
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/config.py:135]: `os.getenv("BUSINESS_EMAILER_TOPIC"`...

File: queue_services/business-pay/src/business_pay/config.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    AUDIENCE = os.getenv(
+        "AUDIENCE", "https://pubsub.googleapis.com/google.pubsub.v1.Subscriber"
+    )
+    PUBLISHER_AUDIENCE = os.getenv(
+        "PUBLISHER_AUDIENCE", "https://pubsub.googleapis.com/google.pubsub.v1.Publisher"
+    )
+    SUB_AUDIENCE = os.getenv("SUB_AUDIENCE", "")
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/config.py:142]: we're missing a secret for this in 1pass I think

File: queue_services/business-pay/src/business_pay/config.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    SUB_SERVICE_ACCOUNT = os.getenv("SUB_SERVICE_ACCOUNT", "")
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/config.py:143]: `AUTHPAY_SERVICE_ACCOUNT` I think? it's calling into business pay

File: queue_services/business-pay/src/business_pay/database/filings.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+                        WHERE f.payment_id = :pay_token
+            """
+            )
+            stmt = stmt.bindparams(pay_token=pay_token)
+
+            filing = Filing.query.from_statement(stmt).one_or_none()
+
+            return filing
+        except Exception as e:
+            print(e)
```
Comment (seeker25) [inline queue_services/business-pay/src/business_pay/database/filings.py:123]: logger?

File: queue_services/business-pay/src/business_pay/config.py (hunk 9 above)
Comment (seeker25) [reply queue_services/business-pay/src/business_pay/config.py:142]: Need a new section similar to `authpay` / `namex` sections in 1 pass

File: queue_services/business-pay/src/business_pay/config.py (hunk 7 above)
Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/config.py:134]: right, I clean that out!

File: queue_services/business-pay/src/business_pay/config.py (hunk 9 above)
Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/config.py:142]: I copied from namex on that one, was hoping ti was already in?

**Review by seeker25 — APPROVED** [review]

File: queue_services/business-pay/src/business_pay/config.py (hunk 9 above)
Comment (seeker25) [reply queue_services/business-pay/src/business_pay/config.py:142]: when the audience is blank in GCP, it's always the destination

File: queue_services/business-pay/src/business_pay/database/filings.py (hunk 11 above)
Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/database/filings.py:123]: yes, missed in my cleanup

File: queue_services/business-pay/devops/vaults.json (hunk 6 above)
Comment (seeker25) [reply queue_services/business-pay/devops/vaults.json:16]: need the new section in here for business.. which has the audience endpoints as well

File: queue_services/business-pay/src/business_pay/config.py (hunk 8 above)
Comment (thorwolpert) [reply queue_services/business-pay/src/business_pay/config.py:135]: not using that yet either, so its out

**Pull request merged**